src/cli/global/upgrade.rs

```diff
@@ -59,56 +61,77 @@ pub(super) async fn upgrade_packages(
     cli_channels: &[String],
     platform: &Platform,
 ) -> miette::Result<()> {
-    // Get channels and versions of globally installed packages
-    let mut installed_versions = HashMap::with_capacity(specs.len());
-    let mut channels = config.compute_channels(cli_channels).into_diagnostic()?;
-
-    for package_name in specs.keys() {
-        let prefix_record = find_installed_package(package_name).await?;
-        let last_installed_channel = Channel::from_str(
-            prefix_record.repodata_record.channel.clone(),
-            config.channel_config(),
-        )
-        .into_diagnostic()?;
-
-        channels.push(last_installed_channel);
-
-        let installed_version = prefix_record
-            .repodata_record
-            .package_record
-            .version
-            .into_version();
-        installed_versions.insert(package_name.clone(), installed_version);
+    let channel_cli = config.compute_channels(cli_channels).into_diagnostic()?;
+
+    // Get channels and version of globally installed packages in parallel
+    let mut channels = HashMap::with_capacity(specs.len());
+    let mut versions = HashMap::with_capacity(specs.len());
+    let mut set: JoinSet<Result<_, Report>> = JoinSet::new();
+    for package_name in specs.keys().cloned() {
+        let channel_config = config.channel_config().clone();
+        set.spawn(async move {
+            let p = find_installed_package(&package_name).await?;
+            let channel =
+                Channel::from_str(p.repodata_record.channel, &channel_config).into_diagnostic()?;
+            let version = p.repodata_record.package_record.version.into_version();
+            Ok((package_name, channel, version))
+        });
+    }
+    while let Some(data) = set.join_next().await {
+        let (package_name, channel, version) = data.into_diagnostic()??;
+        channels.insert(package_name.clone(), channel);
+        versions.insert(package_name, version);
     }
-    channels = channels.into_iter().unique().collect();
 
-    // Fetch sparse repodata
-    let (authenticated_client, sparse_repodata) =
-        get_client_and_sparse_repodata(&channels, *platform, &config).await?;
+    // Fetch sparse repodata across all channels
+    let all_channels = channels.values().chain(channel_cli.iter()).unique();
+    let (client, repodata) =
+        get_client_and_sparse_repodata(all_channels, *platform, &config).await?;
```

Comment:
I think this hasn't changed, but I just noticed if you installed a package from a channel that isn't available anymore, this fails. And it doesn't update any environment. 

Might be nice to filter these and just list which where skipped in the end. 

I'm not avoiding this to merge because of it but just picking your brain. 

The case for me was that I installed a locally build package which I had push to `/tmp` and installed from there.